File: leetcode-everyday/搜索算法/回溯法/46-全排列/46-全排列.cpp
class Solution {
private:
    void dfs(vector<int>& nums, vector<bool>& visited, vector<int>& path, vector<vector<int>>& ans) {
        if(path.size() == nums.size()) {
            ans.push_back(path);
            return;
        }
        for(int i = 0; i < nums.size(); ++i) {
            if(visited[i]) continue;
            visited[i] = true;
            path.push_back(nums[i]);
            dfs(nums, visited, path, ans);
            visited[i] = false;
            path.pop_back();
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        if(nums.empty()) return ans;
        vector<bool> visited(nums.size(), false);
        vector<int> path;
        dfs(nums, visited, path, ans);
        return ans;
    }
};

File: leetcode-everyday/搜索算法/回溯法/46-全排列/46-全排列.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 var permute = function(nums) {
  const ans = [];
  if(!nums.length) return ans;
  const visited = new Array(nums.length).fill(false); 
  dfs(nums, visited, [], ans);
  return ans;
};

function dfs(nums, visited, path, ans) {
  if(path.length === nums.length) {
      ans.push([...path]);
  }
  for(let i = 0; i < nums.length; i++) {
      if(visited[i]) continue;
      visited[i] = true;
      path.push(nums[i]);
      dfs(nums, visited, path, ans);
      visited[i] = false;
      path.pop();
  }
}